### ISSUE TITLE
Memory requirements for running the code

### ISSUE THREAD

**Issue by aayushee** [ameyagodbole/multihop_inference_explanation_regeneration#2]
Hi
I am trying to run bert_reranker.py on Google Colab but the features are not fitting in the 12GB RAM  due to which no further dataset processing is happening. Even if the features file gets cached, loading it again leads to memory crash. Could you please let me know what would be the memory requirements for running the code?
Also, any tips on avoiding loading complete feature file in memory would be greatly appreciated.
Thanks!

**Comment by ameyagodbole** [comment]
We were using a local cluster so RAM wasn't a constraint. That being said we used pickle because it was easy and not because it was efficient.

I can think of a couple of ways to get around the memory issue:
1. Using HDF5 files instead of pickle: They are read directly off of the disk and not loaded into the RAM (at least not entirely).
2. Use some other file format that can be read efficiently. HDF5 is the only alternative that I know of.
3. Creating the features on the fly (There are a couple of ways you could do this)
    - Using PyTorch Dataloaders: You would need to write the appropriate classes but it would handle the multiprocessing for you.
    - Using `yield` instead of [this](https://github.com/ameyagodbole/multihop_inference_explanation_regeneration/blob/ab742433034b251a819b6eb898686530bd055cd0/bert_reranker.py#L351) to return features per example and batch them
    - Downsampling the negative examples: There are significantly more negative labels than the positive ones. You can downsample the negatives and then store the pickle file. We had tried something similar where we trained a model with positive and negative examples sampled equally during training and that model performed only slightly worse than the one we presented.

**Comment by aayushee** [comment]
Thanks for your clarification and suggestions!
